File: office_convert/rate_limit.py
"""Per-client token-bucket rate limiter.

In-memory rate limiting keyed by client IP. Token bucket lets short bursts
through while bounding sustained rate. State is per-process — multi-replica
deployments get N x the configured per-IP rate. Acceptable trade-off until
the service grows beyond a handful of replicas.

Memory is bounded via LRU eviction at `max_keys` entries.

Client identity precedence:
1. First IP in `X-Forwarded-For` (set by ALB/proxy), when `trust_xff` is True.
2. `request.client.host` fallback (direct connection).

NOTE: `X-Forwarded-For` is spoofable when the API is exposed directly. Deploy
behind a proxy that overwrites the header (ALB does), or set
`OFFICE_CONVERT_RATE_LIMIT_TRUST_XFF=0` if exposed directly.
"""

from __future__ import annotations

import asyncio
import time
from collections import OrderedDict
from dataclasses import dataclass

from fastapi import Request


@dataclass
class RateLimitDecision:
    """Result of a rate-limit check. `allowed` says whether to serve."""

    allowed: bool
    limit: int  # tokens-per-minute configured ceiling
    remaining: int  # whole tokens left in the bucket post-decision
    reset_epoch_seconds: int  # epoch seconds when bucket will be full again
    retry_after_seconds: int  # seconds until at least one token is available (>=1 when denied)


class RateLimiter:
    """Token bucket per client identifier.

    Refill rate = `per_minute / 60` tokens/sec. Bucket capacity = `burst`.
    Each request costs 1 token. When the bucket is empty, the request is
    denied with a `retry_after` derived from the refill rate.
    """

    def __init__(self, *, per_minute: int, burst: int, max_keys: int) -> None:
        if per_minute < 1:
            raise ValueError("per_minute must be >= 1")
        if burst < 1:
            raise ValueError("burst must be >= 1")
        if max_keys < 1:
            raise ValueError("max_keys must be >= 1")
        self.per_minute = per_minute
        self.burst = burst
        self.max_keys = max_keys
        self._refill_per_sec = per_minute / 60.0
        # OrderedDict acts as an LRU: move_to_end on touch, popitem(last=False)
        # to evict the oldest. Stores (tokens, last_refill_time_monotonic).
        self._buckets: OrderedDict[str, tuple[float, float]] = OrderedDict()
        self._lock = asyncio.Lock()

    async def check(self, client_id: str) -> RateLimitDecision:
        """Atomically refill the client's bucket and try to consume 1 token."""
        async with self._lock:
            now = time.monotonic()
            if client_id in self._buckets:
                tokens, last = self._buckets[client_id]
                # Refill since last seen.
                tokens = min(float(self.burst), tokens + (now - last) * self._refill_per_sec)
                self._buckets.move_to_end(client_id)
            else:
                tokens = float(self.burst)
                if len(self._buckets) >= self.max_keys:
                    self._buckets.popitem(last=False)

            allowed = tokens >= 1.0
            if allowed:
                tokens -= 1.0
            self._buckets[client_id] = (tokens, now)

            # Time to refill to full bucket from current tokens.
            seconds_to_full = (self.burst - tokens) / self._refill_per_sec
            reset_epoch = int(time.time() + seconds_to_full)

            if allowed:
                retry_after = 0
            else:
                # Always >= 1 — round up so caller gets a positive wait.
                seconds_to_one = (1.0 - tokens) / self._refill_per_sec
                retry_after = max(1, int(seconds_to_one) + 1)

            return RateLimitDecision(
                allowed=allowed,
                limit=self.per_minute,
                remaining=int(tokens),
                reset_epoch_seconds=reset_epoch,
                retry_after_seconds=retry_after,
            )


def client_id_for(request: Request, *, trust_xff: bool) -> str:
    """Resolve a client identifier for rate-limiting.

    When `trust_xff` is True, the first IP in `X-Forwarded-For` wins. This
    matches AWS ALB's behavior of appending the original client IP as the
    first entry. Falls back to `request.client.host`.
    """
    if trust_xff:
        xff = request.headers.get("x-forwarded-for")
        if xff:
            first = xff.split(",", 1)[0].strip()
            if first:
                return first
    if request.client is not None:
        return request.client.host
    return "unknown"

File: internal/ratelimit/ratelimit.go
// Package ratelimit is a per-client token-bucket rate limiter.
//
// Ported from office_convert/rate_limit.py. In-memory, keyed by client IP,
// LRU-bounded at maxKeys. State is per-process (multi-replica gets N× the rate).
package ratelimit

import (
	"container/list"
	"net/http"
	"strings"
	"sync"
	"time"
)

// Decision is the result of a rate-limit check.
type Decision struct {
	Allowed           bool
	Limit             int // tokens-per-minute ceiling
	Remaining         int // whole tokens left post-decision
	ResetEpochSeconds int64
	RetryAfterSeconds int // >=1 when denied
}

type bucket struct {
	tokens float64
	last   time.Time
}

// Limiter is a token bucket per client identifier. Refill rate = perMinute/60
// tokens/sec; capacity = burst; each request costs 1 token.
type Limiter struct {
	perMinute    int
	burst        int
	maxKeys      int
	refillPerSec float64

	mu      sync.Mutex
	buckets map[string]*list.Element // value: *lruEntry
	lru     *list.List
}

type lruEntry struct {
	key string
	b   bucket
}

// New constructs a Limiter. Panics on invalid params (matches Python's
// ValueError at construction).
func New(perMinute, burst, maxKeys int) *Limiter {
	if perMinute < 1 || burst < 1 || maxKeys < 1 {
		panic("ratelimit: per_minute/burst/max_keys must be >= 1")
	}
	return &Limiter{
		perMinute:    perMinute,
		burst:        burst,
		maxKeys:      maxKeys,
		refillPerSec: float64(perMinute) / 60.0,
		buckets:      make(map[string]*list.Element),
		lru:          list.New(),
	}
}

// Check atomically refills the client's bucket and tries to consume 1 token.
func (l *Limiter) Check(clientID string) Decision {
	l.mu.Lock()
	defer l.mu.Unlock()
	now := time.Now()

	var b bucket
	if el, ok := l.buckets[clientID]; ok {
		ent := el.Value.(*lruEntry)
		b = ent.b
		b.tokens = minF(float64(l.burst), b.tokens+now.Sub(b.last).Seconds()*l.refillPerSec)
		l.lru.MoveToBack(el)
	} else {
		b = bucket{tokens: float64(l.burst)}
		if len(l.buckets) >= l.maxKeys {
			if oldest := l.lru.Front(); oldest != nil {
				old := oldest.Value.(*lruEntry)
				delete(l.buckets, old.key)
				l.lru.Remove(oldest)
			}
		}
	}

	allowed := b.tokens >= 1.0
	if allowed {
		b.tokens -= 1.0
	}
	b.last = now

	if el, ok := l.buckets[clientID]; ok {
		el.Value.(*lruEntry).b = b
	} else {
		l.buckets[clientID] = l.lru.PushBack(&lruEntry{key: clientID, b: b})
	}

	secondsToFull := (float64(l.burst) - b.tokens) / l.refillPerSec
	resetEpoch := time.Now().Add(time.Duration(secondsToFull * float64(time.Second))).Unix()

	retryAfter := 0
	if !allowed {
		secondsToOne := (1.0 - b.tokens) / l.refillPerSec
		retryAfter = int(secondsToOne) + 1
		if retryAfter < 1 {
			retryAfter = 1
		}
	}
	return Decision{
		Allowed:           allowed,
		Limit:             l.perMinute,
		Remaining:         int(b.tokens),
		ResetEpochSeconds: resetEpoch,
		RetryAfterSeconds: retryAfter,
	}
}

// ClientIDFor resolves a client identifier. With trustXFF, the first IP in
// X-Forwarded-For wins (ALB appends the original client IP first), else falls
// back to the remote address. Mirrors client_id_for.
func ClientIDFor(r *http.Request, trustXFF bool) string {
	if trustXFF {
		if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
			first := strings.TrimSpace(strings.SplitN(xff, ",", 2)[0])
			if first != "" {
				return first
			}
		}
	}
	host := r.RemoteAddr
	if i := strings.LastIndex(host, ":"); i >= 0 {
		host = host[:i] // strip port
	}
	if host == "" {
		return "unknown"
	}
	return host
}

func minF(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}
